client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java

```diff
@@ -1671,15 +1671,19 @@ protected final ElasticsearchStatusException parseResponseException(ResponseExce
         Response response = responseException.getResponse();
         HttpEntity entity = response.getEntity();
         ElasticsearchStatusException elasticsearchException;
-        if (entity == null) {
+        RestStatus restStatus = RestStatus.fromCode(response.getStatusLine().getStatusCode());
+
+        if (responseException instanceof WarningFailureException) {
+            elasticsearchException = new ElasticsearchStatusException("Warnings/Deprecations caused response to fail", restStatus,
+                responseException);
```

Comment:
I find using ElasticsearchStatusException almost as weird as using ResponseException. That is because such exceptions are meant for error status codes, while warnings can come back with any response.

I would prefer if we could have a separate catch for the warnings case, with a different exception that does not extend `ResponseException`. Maybe we could add a new exception to the low-level client that does not extend ResponseException, and let the high-level client bubble it up as-is? Not sure what I may be missing that prevents us from going down this route...

Reply:
you are right Nik, probably we do not even need a catch for the new exception, just let it bubble up.

Reply:
The only reason I used the existing `ElasticsearchStatusException` was because that is what was thrown before. If you take a look at https://github.com/elastic/elasticsearch/blob/master/client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java#L1670, it is the method that sync/async both call when a `ResponseException` is thrown from the LLRC. Im not sure there is a ton of value wrapping either of these in `ElasticsearchStatusException`. I think removing this would be a breaking change tho?

Reply:
and to @nik9000 point, if we just throw another exception, it works great for sync, but for async it throws the following

```
UncategorizedExecutionException[Failed execution
]; nested: ExecutionException[org.elasticsearch.client.DeprecationException
```

That said, ofc we can make it do whatever we want. Id still lean toward 1) doing nothing different and still throwing a `ElasticsearchStatusException`, or 2) unwrapping both the `ResponseException` and `WarningFailureException` and throwing them w/o the `ElasticsearchStatusException` (breaking?)

Reply:
> UncategorizedExecutionException

We can handle that IIRC. Catch and rethrow pretty much.

I think with a feature this new that is *likely* pretty rare to use we can probably get away with a breaking change here if the result is cleaner.

Reply:
>  likely pretty rare to use

`strictDeprecationMode` defaults to false, so we are def in agreement that its not likely used. Id be ok with throwing this different, but I think that removing the `ElasticsearchStatusException` wrapped from the `ResponseException` thrown would be a bad breaking change.

Reply:
>  I think that removing the `ElasticsearchStatusException` wrapped from the `ResponseException` thrown would be a bad breaking change.

I think that one has been around much longer, yeah. But i'm just going from memory. We could throw a special deprecation exception instead, right?

Reply:
yea this is what the code does as is, even with it being child of ResponseException, its just currently wrapped in a `ElasticsearchStatusException` as well.

Reply:
@javanna Id like your opinion on whether to break the current functionality, which is to always return an `ElasticsearchStatusException` unless its a random exception we dont know about. Ill gladly throw the `WarningFailureException` w/o wrapping it, just want to make sure thats the correct thing to do WRT compatibility. I think @nik9000 says he is ok with it, correct @nik9000 	?

Reply:
I am fine too with rethrowing WarningFailureException as-is.